flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/BigQueryBaseSink.java

```diff
@@ -18,5 +18,58 @@
 
 import org.apache.flink.api.connector.sink2.Sink;
 
+import com.google.cloud.flink.bigquery.common.config.BigQueryConnectOptions;
+import com.google.cloud.flink.bigquery.sink.serializer.BigQueryProtoSerializer;
+import com.google.cloud.flink.bigquery.sink.serializer.BigQuerySchemaProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 /** Base class for developing a BigQuery sink. */
-abstract class BigQueryBaseSink implements Sink {}
+abstract class BigQueryBaseSink implements Sink {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public static final int MAX_SINK_PARALLELISM = 100;
+
+    final BigQueryConnectOptions connectOptions;
+    final BigQuerySchemaProvider schemaProvider;
+    final BigQueryProtoSerializer serializer;
+    final String tablePath;
+
+    BigQueryBaseSink(BigQuerySinkConfig sinkConfig) {
+        validateSinkConfig(sinkConfig);
+        this.connectOptions = sinkConfig.getConnectOptions();
+        this.schemaProvider = sinkConfig.getSchemaProvider();
+        this.serializer = sinkConfig.getSerializer();
+        this.serializer.init(schemaProvider);
+        this.tablePath =
+                String.format(
+                        "projects/%s/datasets/%s/tables/%s",
+                        connectOptions.getProjectId(),
+                        connectOptions.getDataset(),
+                        connectOptions.getTable());
+    }
+
+    private void validateSinkConfig(BigQuerySinkConfig sinkConfig) {
+        if (sinkConfig.getConnectOptions() == null) {
+            throw new IllegalArgumentException("BigQuery connect options cannot be null");
+        }
+        if (sinkConfig.getSerializer() == null) {
+            throw new IllegalArgumentException("BigQuery serializer cannot be null");
+        }
+        if (sinkConfig.getSchemaProvider() == null) {
+            throw new IllegalArgumentException("BigQuery schema provider cannot be null");
+        }
+    }
+
+    /** Ensures Sink's parallelism does not exceed the allowed maximum when scaling Flink job. */
+    void checkParallelism(int numberOfParallelSubtasks) {
```

Comment:
shouldn't we check parallelism as part of validating the sink config? no need to wait until we try to create writers

Reply:
We've moved the parallelism out of sink config because:
1. it is easy to bypass for the user.
2. scaling up the cluster would require another parallelism check which can only be done when creating new writers.
3. first time writer creation happens during job initialization in the beginning so it would be done at the same time as validating sink config.

Writers are created without write streams so we're not invoking BQ storage APIs at all during initialization.

Reply:
okay that's fine. you can resolve this convo; thanks!